csemlib/models/s20rts.py

```diff
@@ -38,17 +47,108 @@ def read(self):
+                    val += getattr(self, 'l%d' % idx)[n][m] * np.imag(sph_harm(imag, n, l, c))
                     imag += 1
 
         return val
+
+    def eval_point_cloud(self, c, l, r, param):
+        """
+        This returns the linearly interpolated perturbations of s20rts. Careful only points that fall inside
+        of the domain of s20rts are returned.
+        :param c: colatitude
+        :param l: longitude
+        :param r: normalised distance from core in km
+        :param param: param to be returned - currently not used
+        :return c, l, r, vals: Re
+        """
+        pts = np.array((c, l, r)).T
+        s20_lay_norm = self.layers / self.r_earth
+
+        # Sorted array, probably not necessary anymore
+        pts_sorted = np.asarray(sorted(pts, key=lambda pts_entry: pts_entry[2], reverse=True))
+
+        # Initialize arrays to store evaluated points in the correct order
+        vals = np.zeros(0)
+        c = np.zeros(0)
+        l = np.zeros(0)
+        r = np.zeros(0)
+
+        for i in range(len(s20_layers) - 1):
+            upper_rad_norm = s20_lay_norm[i]
+            lower_rad_norm = s20_lay_norm[i+1]
+            upper_rad = s20_layers[i]
+            lower_rad = s20_layers[i+1]
+
+            # Extract chunk for interpolation
+            # Discard everything above chunk
+            if i == 0:
+                chunk = pts_sorted[pts_sorted[:, 2] <= upper_rad_norm + np.finfo(float).eps]
+            else:
+                chunk = pts_sorted[pts_sorted[:, 2] <= upper_rad_norm]
+
+            # Discard everything below chunk
+            if i < len(s20_layers) - 2:
+                chunk = chunk[chunk[:, 2] > lower_rad_norm]
+            else:
+                chunk = chunk[chunk[:, 2] >= lower_rad_norm - np.finfo(float).eps]
+
+            chunk_c, chunk_l, chunk_r = chunk.T
+
+            # Evaluate S20RTS at upper and lower end of chunk, use these to interpolate
+            top_vals = self.eval(chunk_c, chunk_l, upper_rad, 'test')
+            bottom_vals = self.eval(chunk_c, chunk_l, lower_rad, 'test')
+
+            chunk_vals = self.linear_interpolation(bottom_vals, top_vals, lower_rad_norm, upper_rad_norm, chunk_r)
+            vals = np.append(vals, chunk_vals)
+            c = np.append(c, chunk_c)
+            r = np.append(r, chunk_r)
+            l = np.append(l, chunk_l)
+
+        return c, l, r, vals
+
+    def find_layer_idx(self, rad):
+        """
+        Return the index of the s20rts layer for the requested depth
+        :param rad: distance from core in km
+        :return layer index:
+        """
+
+        s20_layers = [6346.63, 6296.63, 6241.64, 6181.14, 6114.57, 6041.34, 5960.79, 5872.18, 5774.69, 5667.44, 5549.46,
+                  5419.68,5276.89, 5119.82, 4947.02, 4756.93, 4547.81, 4317.74, 4064.66, 3786.25, 3479.96]
+
+        if rad < s20_layers[-1] or rad > s20_layers[0]:
+            raise ValueError('Requested layer out of bounds for s20rts')
```

Comment:
Hey man,

Think you can cover these lines? I guess you could pass in a radius that is out of bounds, and ensure that it raises a `ValueError`.